db_test.go

```diff
@@ -1301,3 +1301,62 @@ func TestInitializeHeadTimestamp(t *testing.T) {
 		testutil.Equals(t, int64(15000), db.head.MaxTime())
 	})
 }
+
+func TestNoEmptyBlock(t *testing.T) {
+	db, close := openTestDB(t, nil)
+	defer close()
+	defer db.Close()
+
+	// Test no blocks after compact with empty head.
+	_, err := db.compact()
+	testutil.Ok(t, err)
+	testutil.Equals(t, 0, len(db.blocks))
+
+	// Test no blocks after deleting all samples from head.
+	blockRange := DefaultOptions.BlockRanges[0]
+	label := labels.FromStrings("foo", "bar")
+
+	app := db.Appender()
+	for i := int64(0); i < 3; i++ {
+		_, err := app.Add(label, i*blockRange, 0)
+		testutil.Ok(t, err)
+		_, err = app.Add(label, i*blockRange+1000, 0)
+		testutil.Ok(t, err)
+	}
+	err = app.Commit()
+	testutil.Ok(t, err)
+
+	testutil.Ok(t, db.Delete(math.MinInt64, math.MaxInt64, labels.NewEqualMatcher("foo", "bar")))
+	_, err = db.compact()
+	testutil.Ok(t, err)
+	// No blocks created.
+	testutil.Equals(t, 0, len(db.blocks))
+
+	// Test no blocks remaining after small samples are deleted from disk.
+	app = db.Appender()
+	for i := int64(3); i < 20; i++ {
+		for j := int64(0); j < 10; j++ {
+			_, err := app.Add(label, i*blockRange+j, 0)
+			testutil.Ok(t, err)
+		}
+	}
+	err = app.Commit()
+	testutil.Ok(t, err)
+
+	_, err = db.compact()
+	testutil.Ok(t, err)
+	testutil.Assert(t, len(db.blocks) > 0, "No blocks created")
+
+	testutil.Ok(t, db.Delete(math.MinInt64, math.MaxInt64, labels.NewEqualMatcher("foo", "bar")))
+
+	// Mimicking small part of compaction.
```

Comment:
I am bit confused here. Can you give a bit more details what is happening and what are testing for.

Reply:
So we want to test 2 cases here
1. When head is being compacted to create new block. Here if it results in 0 samples then we write no blocks.
2. (which you asked) When compacting a block or multiple blocks, if it results in an empty block, we should test that. Here I am deleting all samples so that no blocks should remain. While testing, `db.compact()` was not selecting the blocks for compaction (basically `Plan()` in compactor was not selecting those directories). So here I am manually mimicking `Plan()` (selecting all blocks) and calling the compactor on those blocks.

Reply:
yep I understand now. so we need some better comments here than.

Something to describe the workflow without looking at each func implementation.

// Build a list of block paths to pass to the compactor.
...
// The compactor should mark all empty blocks as deletable.
...
// The reload should remove all deletable blocks.

Reply:
Updated the comments, with few more checks.